File: openadmin/spec/stat.py
# SPDX-FileCopyrightText: 2026 OpenAdmin
#
# SPDX-License-Identifier: AGPL-3.0-or-later

from typing import Literal, NotRequired, TypedDict

from .colors import Color
from .http_methods import HttpMethod
from .icons import Icon
from .json_schema import JsonSchema


class StatComponent(TypedDict):
    type: Literal["stat"]
    id: str
    icon: Icon | None
    color: Color | None
    name: str
    description: str | None
    method: HttpMethod
    form: JsonSchema | None
    body: JsonSchema | None
    query: JsonSchema | None


type StatValue = str | int | float | bool | None


class StatResponse(TypedDict):
    value: StatValue
    icon: NotRequired[Icon]
    color: NotRequired[Color]


type Stat = StatValue | StatResponse

File: client/src/schemas/stat.ts
// SPDX-FileCopyrightText: 2026 OpenAdmin
//
// SPDX-License-Identifier: AGPL-3.0-or-later

import { z } from "zod"
import { colorSchema } from "./color"
import { httpMethodSchema } from "./http-method"
import { iconSchema } from "./icon"
import { jsonSchemaSchema } from "./json-schema"

export const statComponentSchema = z.object({
	type: z.literal("stat"),
	id: z.string(),
	icon: iconSchema.nullable(),
	color: colorSchema.nullable(),
	name: z.string(),
	refresh: z.number().nullable(),
	description: z.string().nullable(),
	method: httpMethodSchema,
	form: jsonSchemaSchema.nullable(),
	body: jsonSchemaSchema.nullable(),
	query: jsonSchemaSchema.nullable(),
})

export type StatComponent = z.infer<typeof statComponentSchema>

export const statValueSchema = z.union([z.string(), z.number(), z.boolean(), z.null()])

export type StatValue = z.infer<typeof statValueSchema>

export const statResponseSchema = z.object({
	value: statValueSchema,
	refresh: z.number().nullable().optional(),
	icon: iconSchema.optional(),
	color: colorSchema.optional(),
})

export type StatResponse = z.infer<typeof statResponseSchema>

export const statSchema = z.union([statValueSchema, statResponseSchema])

export type Stat = z.infer<typeof statSchema>
